c10/macros/Macros.h

```diff
@@ -201,6 +205,30 @@ constexpr uint32_t CUDA_THREADS_PER_BLOCK_FALLBACK = 256;
 #define CUDA_KERNEL_ASSERT(...) assert(__VA_ARGS__)
 #endif // __APPLE__
 
+// CUDA_ALWAYS_ASSERT is similar to CUDA_KERNEL_ASSERT but checks the assertion
+// even when NDEBUG is defined. This is useful for important assertions in CUDA
+// code that when building Release.
+#if defined(__APPLE__) || defined(__HIP_PLATFORM_HCC__)
+#define CUDA_ALWAYS_ASSERT(cond)
+#else // __APPLE__
+#if defined(NDEBUG)
+extern "C" {
+  [[noreturn]]
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__) || defined(__HIP__)
+    __host__ __device__
+#endif // __CUDA_ARCH__
+  void __assert_fail(const char *assertion, const char *file,
```

Comment:
In MSVC, the function is named as `_wassert` and you should declare it as the following.
```c++
_declspec(dllimport) void _wassert(
	wchar_t const* _Message,
	wchar_t const* _File,
	unsigned       _Line
);

#define __assert_fail(expression) (void)(                                                 \
        (!!(expression)) ||                                                               \
        (_wassert(_CRT_WIDE(#expression), _CRT_WIDE(__FILE__), (unsigned)(__LINE__)), 0)  \
    )
```
Actually, this is copied from `assert.h` in MSVC.

Reply:
Sorry for putting it on hold for the past week.

Thanks for the hint. Unfortunately, I have no way to test this, and so I would rather not commit it. I propose that I will simply disable CUDA_ALWAYS_ASSERT for MSVC builds, comment with a TODO and create a new issue to implement it later.

The tests should pass nevertheless since the one test affected is already disabled on Windows (`skipIf(IS_WINDOWS, 'FIXME: CUDA OOM error on Windows')`).

Reply:
Ok, no problem for me. I'll fix that for you after this PR goes in.